fl4health/clients/deep_mmd_clients/ditto_deep_mmd_client.py

```diff
@@ -0,0 +1,201 @@
+        self.deep_mmd_losses = {}
+        for layer in self.flatten_feature_extraction_layers.keys():
+            self.deep_mmd_losses[layer] = DeepMmdLoss(
+                device=self.device,
+                input_size=self.size_feature_extraction_layers[layer],
+                layer_name=layer,
+            ).to(self.device)
+
+        self.init_global_model: nn.Module
+        self.local_feature_extractor: FeatureExtractorBuffer
+        self.init_global_feature_extractor: FeatureExtractorBuffer
+
+    def setup_client(self, config: Config) -> None:
+        super().setup_client(config)
+        self.local_feature_extractor = FeatureExtractorBuffer(
+            model=self.model,
+            flatten_feature_extraction_layers=self.flatten_feature_extraction_layers,
+        )
+
+    def update_before_train(self, current_server_round: int) -> None:
+        super().update_before_train(current_server_round)
+        assert isinstance(self.global_model, nn.Module)
+        # Register hooks to extract features from the local model if not already registered
+        self.local_feature_extractor._maybe_register_hooks()
+        # Clone and freeze the initial weights GLOBAL MODEL. These are used to form the Ditto local
+        # update penalty term.
+        self.init_global_model = self.clone_and_freeze_model(self.global_model)
+        self.init_global_feature_extractor = FeatureExtractorBuffer(
+            model=self.init_global_model,
+            flatten_feature_extraction_layers=self.flatten_feature_extraction_layers,
+        )
+        # Register hooks to extract features from the init global model if not already registered
+        self.init_global_feature_extractor._maybe_register_hooks()
+
+    def predict(
+        self,
+        input: TorchInputType,
+    ) -> Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]:
+        """
+        Computes the predictions for both the GLOBAL and LOCAL models and pack them into the prediction dictionary
+
+        Args:
+            input (Union[torch.Tensor, Dict[str, torch.Tensor]]): Inputs to be fed into both models.
+
+        Returns:
+            Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]: A tuple in which the first element
+            contains predictions indexed by name and the second element contains intermediate activations
+            index by name.
+
+        Raises:
+            ValueError: Occurs when something other than a tensor or dict of tensors is returned by the model
+            forward.
+        """
+
+        # We use features from init_global_model to compute the DEEP-MMD loss not the global_model
+        global_preds = self.global_model(input)
+        local_preds = self.model(input)
+        features = self.local_feature_extractor.get_extracted_features()
+        if self.deep_mmd_loss_weight != 0:
+            # Compute the features of the init_global_model
+            _ = self.init_global_model(input)
```

Comment:
I don't think you need to catch this with `_` you're not going to store it anyway.